In [5]:
from sympy import symbols, integrate

# Define the symbols
x, y = symbols('x y')

# Define the function to integrate
f = x**2 - 3*y**2 + x*y**3

# Perform the inner integral over x from 0 to 4
inner_integral = integrate(f, (x, 0, 4))

# Now perform the outer integral over y from -2 to 2
outer_integral = integrate(inner_integral, (y, -2, 2))

inner_integral, outer_integral

(8*y**3 - 12*y**2 + 64/3, 64/3)

### Trapezoidal Rule

In [9]:
# Define the function to integrate
f = x**2 - 3*y**2 + x*y**3

# Calculate delta x and delta y
delta_x = (2 - (-2)) / 2
delta_y = (4 - 0) / 2

# Evaluate the function at the corners
f_corner_1 = f.subs({x: 0, y: -2})
f_corner_2 = f.subs({x: 0, y: 2})
f_corner_3 = f.subs({x: 4, y: -2})
f_corner_4 = f.subs({x: 4, y: 2})

# Evaluate the function at the midpoints of the edges
f_edge_1 = f.subs({x: 2, y: -2})
f_edge_2 = f.subs({x: 2, y: 2})
f_edge_3 = f.subs({x: 0, y: 0})
f_edge_4 = f.subs({x: 4, y: 0})

# Evaluate the function at the interior points
f_inner_1 = f.subs({x: 2, y: 0})

# Calculate the trapezoidal sum
trapezoidal_sum = (delta_x * delta_y / 4) * (
    f_corner_1 + f_corner_2 + f_corner_3 + f_corner_4 +
    2 * (f_edge_1 + f_edge_2 + f_edge_3 + f_edge_4) +
    4 * f_inner_1
)

trapezoidal_sum.evalf()

# Recalculate the values at corners, edges, and interior point for trapezoidal rule
f_corner_1 = f.subs({x: 0, y: -2})
f_corner_2 = f.subs({x: 0, y: 2})
f_corner_3 = f.subs({x: 4, y: -2})
f_corner_4 = f.subs({x: 4, y: 2})

# Recalculate the values at the midpoints of the edges
f_edge_1 = f.subs({x: 2, y: -2})
f_edge_2 = f.subs({x: 2, y: 2})
f_edge_3 = f.subs({x: 0, y: 0})
f_edge_4 = f.subs({x: 4, y: 0})

# Recalculate the value at the interior point
f_inner_1 = f.subs({x: 2, y: 0})

# Values for corners, edges, and interior point
corner_values = [f_corner_1, f_corner_2, f_corner_3, f_corner_4]
edge_values = [f_edge_1, f_edge_2, f_edge_3, f_edge_4]
interior_values = [f_inner_1]

# Calculate the trapezoidal sum using the correct formula
trapezoidal_sum_corrected = (delta_x * delta_y / 4) * (
    sum(corner_values) +
    2 * sum(edge_values) +
    4 * sum(interior_values)
)

corner_values, edge_values, interior_values, trapezoidal_sum_corrected.evalf()

([-12, -12, -28, 36], [-24, 8, 0, 16], [4], 0)

The trapezoidal sum using these values is still calculated to be 0. This result is intriguing because, as mentioned earlier, the function and the limits of integration do not suggest symmetry that would cause the integral to cancel out.

### Simpsons 1/3 Rule

In [7]:
# Calculate delta x and delta y for Simpson's rule
delta_x_simpson = (2 - (-2)) / 2
delta_y_simpson = (4 - 0) / 2

# Evaluate the function at the corners for Simpson's rule
f_corner_s1 = f.subs({x: 0, y: -2})
f_corner_s2 = f.subs({x: 0, y: 2})
f_corner_s3 = f.subs({x: 4, y: -2})
f_corner_s4 = f.subs({x: 4, y: 2})

# Evaluate the function at the midpoints of the edges for Simpson's rule
f_edge_s1 = f.subs({x: 2, y: -2})
f_edge_s2 = f.subs({x: 2, y: 2})
f_edge_s3 = f.subs({x: 0, y: 1})  # Note: y midpoint is (0+4)/2 = 2
f_edge_s4 = f.subs({x: 4, y: 1})  # Note: y midpoint is (0+4)/2 = 2

# Evaluate the function at the center point for Simpson's rule
f_center = f.subs({x: 2, y: 1})  # Note: Center point is (0+4)/2, (-2+2)/2

# Calculate the Simpson's sum using the correct formula
simpsons_sum = (delta_x_simpson * delta_y_simpson / 9) * (
    f_corner_s1 + f_corner_s2 + f_corner_s3 + f_corner_s4 +
    4 * (f_edge_s1 + f_edge_s2 + f_edge_s3 + f_edge_s4) +
    16 * f_center
)

simpsons_sum.evalf()

10.6666666666667

The value of the double integral using a single application of Simpson's 
1
3
3
1
​
  rule is approximately 
10.67
10.67. 